[Feat] #55 Firebase Auth를 이용한 구글 로그인 기능 구현

mnh4140 opened the pull request:
close #55 
<!-- close #No(이슈 넘버 등록하면 자동으로 이슈가 닫힙니다. 이슈를 닫고 싶지 않다면, 즉 해당 이슈의 작업이 아직 끝나지 않았다면 닫지 않습니다.) -->

## *⛳️ Work Description*

<!-- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
### 구글 로그인 기능 구현
#### 구글 로그인 라이브러리 추가
- GoogleSignIn
#### 구글 로그인 기능 코드 파일 추가
- AuthService : 구글 로그인 결과를 Firebase Auth 인증 기능
#### 테스트용 코드 파일 추가
- TestLoginViewController : 테스트용 로그인 뷰컨
- TestLoginViewModel : 테스트용 로그인 뷰 모델
#### Firestore 파일 추가
- FirestoreService : 로그인 사용자 정보 Firestore DB 에 저장하는 기능

## *📸 Screenshot*
![Simulator Screen Recording - iPhone 16 Pro - 2025-06-13 at 02 21 01](https://github.com/user-attachments/assets/82d41570-4e2f-41ac-98b8-4ebfc02c0e34)

<!-- 구현한 부분의 시뮬레이터 스크린샷을 올려주시면 됩니다. -->
<!-- 기기대응을 위해 사이즈가 다른 기기별로 스크린샷을 올리기도 합니다. -->
<!-- 움직이는 동작일 경우, 시뮬레이터 영상을 녹화하고 gif로 저장하여 드래그 앤 드롭을 통해 업로드하면 됩니다. -->

## *📢 To Reviewers*

<!-- 이 PR을 리뷰하는 사람들에게 할 말이 있다면 작성하시면 됩니다. -->

### 먼저 운동 쪽 데이터를 저장하는 기능을 먼저 적용하기 위해 테스트용 로그인 뷰컨에 작업했습니다.
### 실제 로그인 뷰에 적용은 추후에 예정입니다.
- 코드리뷰 부탁드립니다.

## *✅ Checklist*

<!-- PR을 올리기 전에 PR을 올리는 대상자가 점검할 내용입니다. -->

- [x]  주석 및 프린트문 제거 확인.
- [x]  컨벤션 준수 확인.

mnh4140 committed:
[Feat] #55 구글 로그인 라이브러리 추가

mnh4140 committed:
[Feat] #55 구글 로그인 연동 설정

mnh4140 committed:
[Feat] #55 구글 로그인 인증 메서드 추가

mnh4140 committed:
[Feat] #55 Firebase Auth를 통한 구글 로그인 기능 구현

Sophie4han commented on FitMate/FitMate/Common/Service/Firebase/Firestore/FirestoreService.swift:
```diff
@@ -0,0 +1,183 @@
+    
+    /// 코드 중복 검사
+    /// 초대 코드 및 운동 코드로 사용 예정
+    func checkInviteCodeDuplicate(code: String, completion: @escaping (Bool) -> Void) {
+        db.collection("users")
+            .whereField("inviteCode", isEqualTo: code)
+            .getDocuments { snapshot, error in
```
주석으로 메서드 설명 부탁드립니도

tututu08 commented on FitMate/FitMate/Common/Service/Firebase/Auth/AuthService.swift:
```diff
@@ -0,0 +1,74 @@
+        GIDSignIn.sharedInstance.configuration = config
+    }
+
+    /// - 구글 로그인 화면을 띄움
+    /// - 사용자 로그인 성공 시, Firebase 인증 처리
+    /// - 성공한 사용자를 Rx의 Single<User> 로 방출
+    func signInWithGoogle(presentingVC: UIViewController) -> Single<FirebaseAuth.User> {
+        return Single.create { observer in
+            // Swuft Concurrency 에서 비동기 코드를 실행하기 위해 Task 사용
+            Task { @MainActor in // @MainActor를 붙인 이유: 로그인 UI를 띄우는 작업이므로 반드시 메인 스레드에서 실행되어야 함
+                do {
+                    // 구글 로그인 화면을 띄우고, 사용자가 로그인할 때까지 기다림.
+                    // try await: 실패할 수도 있는 비동기 작업이기 때문에 try와 await을 함께 사용
+                    // 로그인 결과(GIDGoogleUser)를 await으로 돌려줌
+                    // await : 비동기 함수의 결과가 올 때까지 잠깐 멈췄다가 다시 이어서 실행하라는 뜻.
+                    let signInResult = try await GIDSignIn.sharedInstance.signIn(withPresenting: presentingVC)
+
+                    // 구글 로그인 결과에서 idToken 을 꺼냄
+                    // idToken은 구글이 "이 유저는 인증됨"을 보증해주는 서명된 문자열
+                    // 값이 없으면, 실패로 간주하고 .failure 이벤트를 내보냄
+                    guard let idToken = signInResult.user.idToken?.tokenString else {
+                        observer(.failure(NSError(domain: "GoogleAuth", code: -2)))
+                        return
+                    }
+
+                    // Firebase 인증용 Credential 만들기
+                    // Firebase에게 “이 사람은 구글에서 인증됐어요”라고 증명하기 위한 자격 증명 (credential)을 생성
```
주석으로 상세히 설명해주셔서 이해하기 좋네요 👍

tututu08 commented on FitMate/FitMate/Common/Service/Firebase/Firestore/FirestoreService.swift:
```diff
@@ -0,0 +1,183 @@
+             onFailure: { error in print("삭제 실패: \(error.localizedDescription)") }
+         )
+         .disposed(by: disposeBag)
+     */
```
오 센스쟁이 허니노훈님🍯  궁금했는데 또 달달하게 적어주셨네요 이해했습니다!

tututu08 commented on FitMate/FitMate/Common/Service/Firebase/Firestore/FirestoreService.swift:
```diff
@@ -0,0 +1,183 @@
+        )
+     */
+    
+    // MARK: - Update
```
오 센스쟁이 허니노훈님🍯 궁금했는데 또 달달하게 적어주셨네요 이해했습니다!

tututu08 commented on FitMate/FitMate/Common/Service/Firebase/Firestore/FirestoreService.swift:
```diff
@@ -0,0 +1,183 @@
+             }
+         )
+         .disposed(by: disposeBag)
+     */
```
오 센스쟁이 허니노훈님🍯 궁금했는데 또 달달하게 적어주셨네요 이해했습니다!

tututu08 commented on FitMate/FitMate/Common/Service/Firebase/Firestore/FirestoreService.swift:
```diff
@@ -0,0 +1,183 @@
+            onSuccess: { print("생성 성공") },
+            onFailure: { error in print("실패: \(error)") }
+        )
+     */
```
오 센스쟁이 허니노훈님🍯 궁금했는데 또 달달하게 적어주셨네요 이해했습니다!

tututu08 commented on FitMate/FitMate/Common/Service/Firebase/Firestore/FirestoreService.swift:
```diff
@@ -0,0 +1,183 @@
+            .getDocuments { snapshot, error in
+                if let error = error {
+                    print("Firestore error: \(error)")
+                    // 에러가 났을 때는 '중복 아님'으로 처리하는 게 안전하지 않으니 false로 처리
```
좋습니다 👍

tututu08 commented on FitMate/FitMate/Common/Service/Firebase/Firestore/FirestoreService.swift:
```diff
@@ -0,0 +1,183 @@
+                    completion(false)
+                    return
+                }
+                // 중복이 있다면 count > 0
+                if let count = snapshot?.documents.count, count > 0 {
+                    completion(true) // 이미 존재(중복)
+                } else {
+                    completion(false) // 중복 아님
+                }
+            }
+    }
```
중복처리까지 담당해주는 허니노훈님 🍯

tututu08 commented on FitMate/FitMate/Common/Service/Firebase/Firestore/FirestoreService.swift:
```diff
@@ -0,0 +1,183 @@
+                                print("User 데이터 생성 실패: \(error.localizedDescription)")
+                            } else {
+                                single(.success(()))
+                                print("User 데이터 생성 완료: \(uid), 초대코드: \(inviteCode)")
```
사용하는 방법까지 보니 흐름이 이해 됩니다 👍

Kimewg commented on FitMate/FitMate/Common/Service/Firebase/Auth/LoginTest/TestLoginViewController.swift:
```diff
@@ -0,0 +1,104 @@
+import RxRelay
+import FirebaseAuth
+
+class TestLoginViewController: UIViewController {
```
나중에 base로 변경해주세여..!

mnh4140 replied:
삭제할 친구라서...

Kimewg commented on FitMate/FitMate/Common/Service/Firebase/Auth/LoginTest/TestLoginViewController.swift:
```diff
@@ -0,0 +1,104 @@
+            .drive(onNext: { [weak self] nav in
+                // ViewModel에서 전달한 목적에 따라 화면 이동만 수행
+                switch nav {
+                case .goToSeleteSport(let uid):
```
알럿 케이스로 나눠서 처리하는 거 굿

Kimewg commented on FitMate/FitMate/Common/Service/Firebase/Firestore/FirestoreService.swift:
```diff
@@ -0,0 +1,183 @@
+    /// 초대 코드 및 운동 코드로 사용 예정
+    func checkInviteCodeDuplicate(code: String, completion: @escaping (Bool) -> Void) {
+        db.collection("users")
+            .whereField("inviteCode", isEqualTo: code)
```
이게 뭔가여..? 특정 문서 찾는건가여?

mnh4140 replied:
사용자 정보가 저장된 DB에서 `inviteCode` 값을 찾는 로직입니다!

Sophie4han commented on FitMate/FitMate/Common/Service/Firebase/Auth/AuthService.swift:
```diff
@@ -0,0 +1,74 @@
+            // Swuft Concurrency 에서 비동기 코드를 실행하기 위해 Task 사용
+            Task { @MainActor in // @MainActor를 붙인 이유: 로그인 UI를 띄우는 작업이므로 반드시 메인 스레드에서 실행되어야 함
+                do {
+                    // 구글 로그인 화면을 띄우고, 사용자가 로그인할 때까지 기다림.
+                    // try await: 실패할 수도 있는 비동기 작업이기 때문에 try와 await을 함께 사용
+                    // 로그인 결과(GIDGoogleUser)를 await으로 돌려줌
+                    // await : 비동기 함수의 결과가 올 때까지 잠깐 멈췄다가 다시 이어서 실행하라는 뜻.
+                    let signInResult = try await GIDSignIn.sharedInstance.signIn(withPresenting: presentingVC)
```
주석 설명으로 이해를 도우신 점 매우 좋습니도

Kimewg commented on FitMate/FitMate/Common/Service/Firebase/Auth/AuthService.swift:
```diff
@@ -0,0 +1,74 @@
+            Task { @MainActor in // @MainActor를 붙인 이유: 로그인 UI를 띄우는 작업이므로 반드시 메인 스레드에서 실행되어야 함
+                do {
+                    // 구글 로그인 화면을 띄우고, 사용자가 로그인할 때까지 기다림.
+                    // try await: 실패할 수도 있는 비동기 작업이기 때문에 try와 await을 함께 사용
```
await는 이럴 때 쓰는거군요?

mnh4140 committed:
[Merge] branch 'dev' into Feat/#55

# Conflicts:
#	FitMate/FitMate/App/SceneDelegate.swift
#	FitMate/FitMate/Common/Service/Firebase/Firestore/FirestoreService.swift
#	FitMate/FitMate/Resources/Info.plist

mnh4140 committed:
[Feat] #55 MainView 에 구글 로그인 기능 연결

Sophie4han reviewed (approved):
수고하셨습니도
